### PR TITLE
[MIG 10] Product quick bom

### PR DESCRIPTION
**Pull request by EBII**

### CODE REVIEW

**Comment by bealdav** [comment]
:+1:

File: product_quick_bom/tests/test_quick_bom.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        hard_drive = self.hard_drive
+        cpu = self.cpu
+        computer.bom_line_ids = [
+            (0, 0, {'product_id': ram.id, 'product_qty': 2, }),
+            (0, 0, {'product_id': hard_drive.id, 'product_qty': 2, }),
+            (0, 0, {'product_id': cpu.id, 'product_qty': 1, })
+        ]
+
+        bom = self.env['mrp.bom'].search([('product_id', '=', computer.id)])
+        self.assertIsNotNone(bom)
```
Comment (sebastienbeau) [inline product_quick_bom/tests/test_quick_bom.py]: Just asserting that the bom exist is light, you can add some check on the bom line.
You can check that we have 3 bom line and that the product and qty are correct

File: product_quick_bom/tests/test_quick_bom.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        self.computer = self.env.ref(
+            'product.product_product_5_product_template')
+        self.ram = self.env.ref('product.product_product_13_product_template')
+        self.hard_drive = self.env.ref(
+            'product.product_product_17_product_template')
+        self.cpu = self.env.ref('product.product_product_22_product_template')
+
+    def test_create_bom(self):
+
+        computer = self.computer
```
Comment (sebastienbeau) [inline product_quick_bom/tests/test_quick_bom.py]: you do not need to set computer = self.computer
you can directly use self.computer

File: product_quick_bom/tests/test_quick_bom.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+            (0, 0, {'product_id': ram.id, 'product_qty': 2, }),
+            (0, 0, {'product_id': hard_drive.id, 'product_qty': 2, }),
+            (0, 0, {'product_id': cpu.id, 'product_qty': 1, })
+        ]
+
+        bom = self.env['mrp.bom'].search([('product_id', '=', computer.id)])
+        self.assertIsNotNone(bom)
+
+    def test_read_bom(self):
+        computer = self.computer
```
Comment (sebastienbeau) [inline product_quick_bom/tests/test_quick_bom.py]: same here use self.compute, self.ram directly

**Review by sebastienbeau — CHANGES_REQUESTED** [review]

**Comment by LoisRForgeFlow** [comment]
This error raises in runbot when saving some bom lines in this product:
image
image

**Comment by EBII** [comment]
i don't know where something overwrite default value with unknown selection

File: product_quick_bom/tests/test_quick_bom.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        self.computer = self.env.ref(
+            'product.product_product_5_product_template')
+        self.ram = self.env.ref('product.product_product_13_product_template')
+        self.hard_drive = self.env.ref(
+            'product.product_product_17_product_template')
+        self.cpu = self.env.ref('product.product_product_22_product_template')
+
+    def test_create_bom(self):
+
+        # computer._process_bom_vals({'bom_line_ids' : [
```
Comment (sebastienbeau) [inline product_quick_bom/tests/test_quick_bom.py]: useless comment

File: product_quick_bom/tests/test_quick_bom.py [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+        #     (0, 0, {'product_id': ram.id, 'product_qty': 2, }),
+        #     (0, 0, {'product_id': hard_drive.id, 'product_qty': 2, }),
+        #     (0, 0, {'product_id': cpu.id, 'product_qty': 1, })] })
+        self.computer.write({'bom_line_ids': [
+            (0, 0, {'product_id': self.ram.id, 'product_qty': 2, }),
+            (0, 0, {'product_id': self.hard_drive.id, 'product_qty': 2, }),
+            (0, 0, {'product_id': self.cpu.id, 'product_qty': 1, })]})
+        bom = self.computer.bom_ids
+        self.assertEqual(self.computer.id, bom.product_tmpl_id.id)
```
Comment (sebastienbeau) [inline product_quick_bom/tests/test_quick_bom.py:27]: you have declared the variable bom you can use it after

File: product_quick_bom/product.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+    @api.multi
+    def write(self, vals):
+        """ default_type is in product and bom so _context ovewrite the default
+                value for bom with typ:'product' need to becarefull
+        """
+        bom_vals = self._extract_bom_line(vals)
+        res = super(ProductTemplate, self).write(vals)
+        if bom_vals:
+            ctx = self._context.copy()
```
Comment (sebastienbeau) [inline product_quick_bom/product.py]: ctx = self._context.copy()
ctx.pop('default_type', None)
self.with_context(ctx)._process_bom_vals(bom_vals)

**Review by sebastienbeau — CHANGES_REQUESTED** [review]

File: product_quick_bom/README.rst [hunk 7]
```diff
@@ -0,0 +1,63 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+   :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+   :alt: License: AGPL-3
+
+=================
+product_quick_bom
```
Comment (LoisRForgeFlow) [inline product_quick_bom/README.rst]: `Product Quick Bom` as in the manifest

File: product_quick_bom/README.rst [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+=====
+
+To use this module, you need to:
+
+* go to a product form, click on edit, add a line in the bom tab and see that a bom
+  linked to the product have been created
+
+.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/manufacture/10.0
```
Comment (LoisRForgeFlow) [inline product_quick_bom/README.rst]: The correct link is: https://runbot.odoo-community.org/runbot/129/10.0

File: product_quick_bom/README.rst [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+Indeed in the product form you now have a new tab "bom" that give you the posibility
+to add the bom line directly here
+
+Usage
+=====
+
+To use this module, you need to:
+
+* go to a product form, click on edit, add a line in the bom tab and see that a bom
+  linked to the product have been created
```
Comment (LoisRForgeFlow) [inline product_quick_bom/README.rst]: Start sentences with high caps and end them with a dot.

File: product_quick_bom/__manifest__.py [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+    "website": "https://odoo-community.org/",
+    "author": "Akretion, Odoo Community Association (OCA)",
+    "license": "AGPL-3",
+    "application": False,
+    "installable": True,
+    "depends": [
+        "mrp",
+    ],
+    "data": [
+        "product_view.xml",
```
Comment (LoisRForgeFlow) [inline product_quick_bom/__manifest__.py]: It would be good to move this to a `views` folder.

File: product_quick_bom/bom.py [hunk 11]
```diff
@@ -0,0 +1,25 @@
+# -*- coding: utf-8 -*-
```
Comment (LoisRForgeFlow) [inline product_quick_bom/bom.py]: Same for models, move them to a `models` folder to follow convention.

File: product_quick_bom/tests/__init__.py [hunk 12]
```diff
@@ -0,0 +1,5 @@
+# -*- coding: utf-8 -*-
+#   Copyright (C) 2015 Akretion (http://www.akretion.com).
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+from . import test_quick_bom
```
Comment (LoisRForgeFlow) [inline product_quick_bom/tests/__init__.py]: add end blank line

**Review by LoisRForgeFlow — CHANGES_REQUESTED** [review]
Functionally works fine, some comments:

**Review by LoisRForgeFlow — APPROVED** [review]
Code review of the latest changes. LGTM now :+1:

**Review by sebastienbeau — APPROVED** [review]

**Comment by chafique-delli** [comment]
:+1:

**Pull request merged**